content/advent-2017/mphf.md

```diff
@@ -0,0 +1,297 @@
+encountered in hash tables (the `map` keyword in Go). Collisions, where two
+input values hash to the same integer, can be an annoyance in hash tables and
+disastrous in cryptography. As long as the set of strings to be hashed is
+larger than the output size of the hash, there is always the possibility of
+collision. For example, imagine a hash function that produces a single byte
+of output. There are 256 possible output values. If I try to hash 257
+strings, at least one of them must collide -- there just aren't enough
+different outputs available. This is known as the [pigeonhole principle](
+https://en.wikipedia.org/wiki/Pigeonhole_principle).
+
+However, if we know the set of keys in advance, we can be more careful. A
+*perfect* hash function can be constructed that maps each of the keys to a
+distinct integer, with no collisions. These functions only work with the
+specific set of keys they were constructed for. Passing an unknown key will
+result a false match or even crash.
+
+A minimal perfect hash function goes one step further. It maps the N keys to
+exactly the integers 0..N-1, with each key getting precisely one value.
+
+We can rank hash functions on a few different criteria: speed to construct,
+speed to evaluate, and space used. Imagine a hash function that stores every
+key in an array, and just walks down the array looking for a match, then
+returns that integer. Obviously this maps each element to a distinct value,
+and is quick to construct. However, both the space required and the execution
+time are not optimal. We can do better.
+
+Let’s start with a very basic implementation. Here’s an example set of keys.
+Let’s pretend they’re commands for some simple network protocol, like NATS or
+Redis. We’ll read 4 bytes from the network and we want to check if we have a
+valid command before dispatching to the appropriate processing loop.
+
+Here are the commands:
+```
+INFO
+CONN
+PUB 
+SUB 
+UNSU
+PING
+PONG
++OK 
+-ERR
+AUTH
+PUSH
+ADD 
+DECR
+SET 
+GET 
+QUIT
+```
+
+I’ve left the spaces after the three letter commands so that each key fits
+into a uint32. In fact, we won't deal with these as strings but we'll turn
+them directly into uint32s.
+
+Here's our first hash function. It takes the uint32 and returns the bottom 4 bits.
+
+```
+func hash1(x uint32) uint32 {
+    return x & 15
+}
+```
+
+And then check for duplicates:
+
+```
+dups := make(map[uint32]bool)
+for _, k := range keys {
+    h := hash1(k)
+    if dups[h] {
+        log.Println("duplicate found")
+    }
+    dups[h] = true
+}
+```
+
+This is very fast, but when we test, half of the keys collide. Changing the
+function to look at the upper 4 bits doesn't work either. The duplicate
+initial letters (PUSH, PUB) and trailing letters (PONG, PING) means we need
+to do more shuffling.
+
+Here’s a second attempt:
+
+```
+const multiplier = 31
+func hash2(x uint32) int {
+   var u32 = (x * multiplier) >> 28
+   return u32 & 15
```

Comment:
The Go playground doesn't like this line, as you're attempting to `&` an `int` and a `uint32`.  I fixed this by casting u32 to an int.

Reply:
Oops, thanks.  I copied this by hand.  Fixed.